### PR TITLE
[interp] Implement CEE_RETHROW properly by allocating a separate exce…

### PR DESCRIPTION
**Pull request by vargaz**
…ption variable for each clause, and not overwriting stack traces on a rethrow.

### CODE REVIEW

**Comment by vargaz** [comment]
build interpreter

File: mono/mini/interp/transform.c [hunk 1]
```diff
@@ -3743,10 +3741,14 @@ generate (MonoMethod *method, RuntimeMethod *rtm, unsigned char *is_bb_start, Mo
 				ip += 2;
 				break;
 #endif
-			case CEE_RETHROW:
+			case CEE_RETHROW: {
+				int clause_index = td.clause_indexes [in_offset];
+				g_assert (clause_index != -1);
 				SIMPLE_OP (td, MINT_RETHROW);
+				ADD_CODE (&td, rtm->exvar_offsets [clause_index]);
```
Comment (lewurm) [inline mono/mini/interp/transform.c]: this needs an update of `CEE_RETHROW` in `mini/interp/mintops.def`

File: mono/mini/interp/interp.c [hunk 2]
```diff
@@ -604,6 +604,16 @@ fill_in_trace (MonoException *exception, MonoInvocation *frame)
...
+#define THROW_EX_GENERAL(exception,ex_ip,rethrow)		\
+	do {\
+		frame->ip = (ex_ip);		\
+		frame->ex = (MonoException*)(exception);	\
+		if (!rethrow) { \
+			FILL_IN_TRACE(frame->ex, frame);	\
+		} \
+		goto handle_exception;	\
+	} while (0)
+
```
Comment (lewurm) [inline mono/mini/interp/interp.c]: let's define `THROW_EX` in terms of `THROW_EX_GENERAL` then?

File: mono/mini/interp/transform.c [hunk 3]
```diff
@@ -4160,6 +4165,13 @@ mono_interp_transform_method (RuntimeMethod *runtime_method, ThreadContext *cont
 		offset += size;
 	}
 	offset = (offset + 7) & ~7;
+
+	runtime_method->exvar_offsets = g_malloc (header->num_clauses * sizeof (guint32));
+	for (i = 0; i < header->num_clauses; i++) {
+		offset += sizeof (MonoObject*);
+		runtime_method->exvar_offsets [i] = offset;
+	}
```
Comment (lewurm) [inline mono/mini/interp/transform.c]: does this have to be in the storage for `locals` (i.e. is that some specification thing) or should we keep a separate storage for those?

Comment (vargaz) [inline mono/mini/interp/transform.c]: 'locals' seems like a good place for it.

Comment (lewurm) [inline mono/mini/interp/transform.c]: ok.  I just noticed we align `offset` above, I think we should do that after computing the slots for `exvar_offsets`.

File: mono/mini/interp/mintops.def [hunk 4]
```diff
@@ -171,7 +171,7 @@ OPDEF(MINT_BR_S, "br.s", 2, MintOpShortBranch)
 OPDEF(MINT_LEAVE_S, "leave.s", 2, MintOpShortBranch)
 
 OPDEF(MINT_THROW, "throw", 1, MintOpNoArgs)
-OPDEF(MINT_RETHROW, "rethrow", 1, MintOpNoArgs)
+OPDEF(MINT_RETHROW, "rethrow", 1, MintOpUShortInt)
```
Comment (lewurm) [inline mono/mini/interp/mintops.def]: it's length 2 now

**Comment by lewurm** [comment]
build interpreter

**Pull request merged**